community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/MemoryManager.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.io.pagecache.impl.muninn;
+
+import org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil;
+
+/**
+ * The memory manager is simple: it only allocates memory, until it itself is finalizable and frees it all in one go.
+ *
+ * The memory is allocated in large segments, and the memory returned by the memory manager is page aligned, and plays
+ * well with transparent huge pages and other operating system optimisations.
+ *
+ * The memory manager assumes that the memory claimed from it is evenly divisible in units of pages.
+ */
+final class MemoryManager
+{
+    private static final long GRAB_SIZE = 32 * 1024 * 1024; // 32 MiB
+
+    /**
+     * The amount of memory that this memory manager can still allocate.
+     */
+    private long memoryReserve;
+    private final long alignment;
+
+    private Slab slabs;
+
+    /**
+     * Create a new MemoryManager that will allocate the given amount of memory, to pointers that are aligned to the
+     * given alignment size.
+     * @param expectedMaxMemory The maximum amount of memory that this memory manager is expected to allocate. The
+     * actual amount of memory used can end up greater than this value, if some of it gets wasted on alignment padding.
+     * @param alignment The byte multiple that the allocated pointers have to be aligned at.
+     */
+    public MemoryManager( long expectedMaxMemory, long alignment )
+    {
+        this.memoryReserve = expectedMaxMemory;
+        this.alignment = alignment;
+    }
+
+    /**
+     * Allocate a contiguous, aligned region of memory of the given size in bytes.
+     * @param bytes the number of bytes to allocate.
+     * @return A pointer to the allocated memory.
+     */
+    public synchronized long allocateAligned( long bytes )
+    {
+        if ( slabs == null || !slabs.canAllocate( bytes ) )
+        {
+            long slabGrab = Math.min( GRAB_SIZE, memoryReserve );
```

Comment:
No, hang on. We are adding the alignment to _bytes_ and this sum overrides whatever slabGrab was before. The memoryReserver will go negative, and that's fine, because we don't use it in the calculations if it's less than bytes.